### PR TITLE
core: add nonce field to transaction

### LINKED ISSUE #840: Add nonce field to Transaction
It should work the same way as in C# node (it doesn't have a lot of functionality other than serialization/deserialization/JSON and generation in wallet).

### PR DESCRIPTION
**Pull request by AnnaShaleva**
closes #840

### CODE REVIEW

File: pkg/core/transaction/transaction.go [hunk 1]
```diff
@@ -107,6 +110,7 @@ func (t *Transaction) AddVerificationHash(addr util.Uint160) {
 func (t *Transaction) DecodeBinary(br *io.BinReader) {
 	t.Type = TXType(br.ReadB())
 	t.Version = uint8(br.ReadB())
+	t.Nonce = uint32(br.ReadB())
```
Comment (fyrchik) [inline pkg/core/transaction/transaction.go:113]: `Nonce` is 4-bytes, while `ReadB()` reads one.
You should use `br.ReadU32LE` (or `BE`) as in reference implementation.

File: pkg/wallet/account.go [hunk 2]
```diff
@@ -128,6 +129,9 @@ func NewAccount() (*Account, error) {
 
 // SignTx signs transaction t and updates it's Witnesses.
 func (a *Account) SignTx(t *transaction.Transaction) error {
+	if t.Nonce == 0 {
+		t.Nonce = rand.Uint32()
```
Comment (fyrchik) [inline pkg/wallet/account.go:133]: I don't think `SignTx` is a right place for generating nonce. Maybe we could do this while creating transaction?

Comment (AnnaShaleva) [reply pkg/wallet/account.go]: We don't have separate NewTransaction method, so I added nonce generation to `NewContractTx()`, `NewInvocationTx()`, ... and `claimGas`, `transferAsset` methods of `cli/wallet/wallet.go`. However, we have a lot of places, where we use `transaction.Transaction{...}`, should it be added to all of these places?

File: pkg/core/helper_test.go [hunk 3]
```diff
@@ -388,15 +405,36 @@ func TestCreateBasicChain(t *testing.T) {
 		}
 	}
 
+	// Make a NEO roundtrip (send to myself) and claim GAS.
```
Comment (AnnaShaleva) [inline pkg/core/helper_test.go:408]: @roman-khimov

File: pkg/core/helper_test.go [hunk 4]
```diff
@@ -388,15 +405,36 @@ func TestCreateBasicChain(t *testing.T) {
...
+	txNeoRound.AddOutput(&transaction.Output{
+		AssetID:    GoverningTokenID(),
+		Amount:     neoAmount - util.Fixed8FromInt64(1000),
+		ScriptHash: priv0.GetScriptHash(),
+		Position:   0,
+	})
+	txNeoRound.Data = new(transaction.ContractTX)
+	require.NoError(t, acc0.SignTx(txNeoRound))
+	bw := io.NewBufBinWriter()
+	txNeo0to1.EncodeBinary(bw.BinWriter)
```
Comment (roman-khimov) [inline pkg/core/helper_test.go:424]: `s/txNeo0to1/txNeoRound/`?

File: pkg/consensus/consensus_test.go [hunk 5]
```diff
@@ -34,18 +31,18 @@ func TestNewService(t *testing.T) {
...
 	srv := newTestService(t)
 	txs := []*transaction.Transaction{
-		newMinerTx(1),
-		newMinerTx(2),
-		newMinerTx(3),
-		newMinerTx(4),
+		transaction.NewMinerTX(),
+		transaction.NewMinerTX(),
+		transaction.NewMinerTX(),
+		transaction.NewMinerTX(),
```
Comment (roman-khimov) [inline pkg/consensus/consensus_test.go:37]: Shouldn't it be `NewMinerTXWithNonce()`?

Comment (roman-khimov) [reply pkg/consensus/consensus_test.go:37]: And everywhere else in this file. Random nonce might lead to random test results here if there to occur some duplicate nonces.

File: pkg/core/block/block_test.go [hunk 6]
```diff
@@ -104,171 +103,175 @@ func TestBlockVerify(t *testing.T) {
...
-		"dfa5f84366cf0b48f1b1e9b24a73557e657f6ac21b676528401f5a630aece571": false,
-		"5839fbcbbca68aef41dfa9a371222565519626affad6be0977d38a82259480a5": false,
-		"6873568cae35e4ce0a7d07ef080ef6eb699b2b9dcbc419fad1c4f645ff8579fc": false,
-		"dbb3c0688003bede7e7bc56d2c9d6362b594512ac686820739d963ef91e2eb9e": false,
-		"3d12353cb8bae8be928131580e960a82f37ca3ad6957ad22c8cadc1b21b2dd1a": false,
-		"8ca87fd5843f000939244151ce027bad5c1f30f1867c7054918b7f9a66b949e8": false,
-		"ad088940e45a73e00a3cdb7f3248c67a3f6e5d1f05d4cfd44c4e1f4d26cfef87": false,
-		"908a398dd65dfd2aad6c06090c5a71d5e5280746577a6ddd5a1f2c1453f71ead": false,
-	}
+	//TODO NEO3.0: Update binary
```
Comment (roman-khimov) [inline pkg/core/block/block_test.go:127]: If you're to comment out whole function like
```
/*
func ... {
}
*/
```
you won't have this massive useless diff.

File: pkg/core/helper_test.go [hunk 7]
```diff
@@ -33,6 +33,15 @@ var privNetKeys = []string{
 	"L2oEXKRAAMiPEZukwR5ho2S6SMeQLhcK9mF71ZnF7GvT8dU4Kkgz",
 }
 
+// To be incremented after each created transaction to keep chain constant.
+var testNonce uint32 = 1
+
+// Use as nonce when new transaction is created to avoid random data in tests.
+func getNextNonce() uint32 {
```
Comment (roman-khimov) [inline pkg/core/helper_test.go:40]: You can make it internal to `TestCreateBasicChain`.

File: pkg/core/block/block_test.go [hunk 8]
```diff
@@ -104,27 +103,30 @@ func TestBlockVerify(t *testing.T) {
...
+		transaction.NewIssueTX(),
+		transaction.NewIssueTX(),
 	}
 	assert.NotNil(t, block.Verify())
 }
 
+//TODO NEO3.0: Update binary
+/*
 func TestBinBlockDecodeEncode(t *testing.T) {
+
```
Comment (roman-khimov) [inline pkg/core/block/block_test.go:129]: blank line

File: pkg/core/block/block_test.go [hunk 9]
```diff
@@ -191,6 +193,7 @@ func TestBinBlockDecodeEncode(t *testing.T) {
 }
 
 func TestBlockSizeCalculation(t *testing.T) {
+
```
Comment (roman-khimov) [inline pkg/core/block/block_test.go:196]: blank line

File: pkg/core/transaction/enrollment_test.go [hunk 10]
```diff
@@ -1,14 +1,9 @@
...
-	"testing"
-
-	"github.com/nspcc-dev/neo-go/pkg/internal/testserdes"
-	"github.com/stretchr/testify/assert"
-)
-
+//TODO NEO3.0: Update bynary
+/*
 func TestEncodeDecodeEnrollment(t *testing.T) {
+
```
Comment (roman-khimov) [inline pkg/core/transaction/enrollment_test.go:6]: useless blank line.

**Pull request merged**